Go language module crashes on embedded interfaces

The following code from Clouditor crashes the Go parsing lib:

```golang
type Authorizer interface {
	credentials.PerRPCCredentials
	oauth2.TokenSource
}
```

```golang
panic: runtime error: index out of range [0] with length 0
goroutine 17 [running, locked to thread]:
cpg/frontend.(*GoLanguageFrontend).handleInterfaceTypeSpec(0x14000018110?, 0x1246c5c87?, 0x140000221c0, 0x1400000e3a8)
	/Users/oxisto/Repositories/cpg/cpg-language-go/src/main/golang/frontend/handler.go:473 +0x278
cpg/frontend.(*GoLanguageFrontend).handleTypeSpec(0x0?, 0x14000206720?, 0x140000221c0)
	/Users/oxisto/Repositories/cpg/cpg-language-go/src/main/golang/frontend/handler.go:378 +0x118
cpg/frontend.(*GoLanguageFrontend).handleGenDecl(0x0?, 0x1246c28c3?, 0x1e?)
	/Users/oxisto/Repositories/cpg/cpg-language-go/src/main/golang/frontend/handler.go:327 +0x19c
cpg/frontend.(*GoLanguageFrontend).handleDecl(0x1246bdddc?, 0x14000022040?, {0x124723a38?, 0x14000022280?})
	/Users/oxisto/Repositories/cpg/cpg-language-go/src/main/golang/frontend/handler.go:167 +0xc4
cpg/frontend.(*GoLanguageFrontend).HandleFile(0x1400014fbf8?, 0x1400001e080?, 0x14000134300, {0x1400001e580, 0x36})
	/Users/oxisto/Repositories/cpg/cpg-language-go/src/main/golang/frontend/handler.go:114 +0x490
main.Java_de_fraunhofer_aisec_cpg_frontends_golang_GoLanguageFrontend_parseInternal(0x12dd0cf28, 0x1718aa248, 0x1718aa240, 0x1718aa238, 0x1718aa230)
	/Users/oxisto/Repositories/cpg/cpg-language-go/src/main/golang/lib/cpg/main.go:103 +0x484
```

The handler tries to parse this as a method (which it isn't). The closest we could do here is parse the interfaces similar to a Java `extends`.

--- a/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/NodeBuilder.kt
+++ b/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/NodeBuilder.kt
@@ -931,8 +931,7 @@ object NodeBuilder {
         lang: LanguageFrontend? = null,
         rawNode: Any? = null
     ): ProblemExpression {
-        val node = ProblemExpression()
-        node.problem = problem
+        val node = ProblemExpression(problem, type)
         node.setCodeAndRegion(lang, rawNode, code)
         log(node)
         return node

--- a/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/declarations/RecordDeclaration.java
+++ b/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/declarations/RecordDeclaration.java
@@ -253,6 +253,15 @@ public class RecordDeclaration extends Declaration implements DeclarationHolder,
   }
 
   /**
+   * Adds a type to the list of super classes for this record declaration.
+   *
+   * @param superClass the super class.
+   */
+  public void addSuperClass(Type superClass) {
+    this.superClasses.add(superClass);
+  }
+
+  /**
    * Interfaces implemented by this class. This concept is not present in C++
    *
    * @return the list of implemented interfaces

--- a/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/statements/expressions/ProblemExpression.kt
+++ b/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/statements/expressions/ProblemExpression.kt
@@ -44,15 +44,15 @@ class ProblemExpression(
             .toString()
     }
 
-    override fun equals(o: Any?): Boolean {
-        if (this === o) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) {
             return true
         }
-        if (o !is ProblemExpression) {
+        if (other !is ProblemExpression) {
             return false
         }
-        val that = o
-        return (super.equals(that) && problem == that.problem)
+
+        return (super.equals(other) && problem == other.problem)
     }
 
     override fun hashCode(): Int {

--- a/cpg-language-go/src/main/golang/declarations.go
+++ b/cpg-language-go/src/main/golang/declarations.go
@@ -162,6 +162,12 @@ func (r *RecordDeclaration) AddMethod(m *MethodDeclaration) (err error) {
 	return
 }
 
+func (r *RecordDeclaration) AddSuperClass(t *Type) (err error) {
+	(*jnigi.ObjectRef)(r).CallMethod(env, "addSuperClass", nil, t)
+
+	return
+}
+
 func (r *RecordDeclaration) IsNil() bool {
 	return (*jnigi.ObjectRef)(r).IsNil()
 }

--- a/cpg-language-go/src/main/golang/expressions.go
+++ b/cpg-language-go/src/main/golang/expressions.go
@@ -49,6 +49,7 @@ func (e *Expression) IsArray() bool {
 }
 
 type CallExpression Expression
+type CastExpression Expression
 type NewExpression Expression
 type ArrayCreationExpression Expression
 type ArraySubscriptionExpression Expression
@@ -73,6 +74,19 @@ func NewCallExpression(fset *token.FileSet, astNode ast.Node) *CallExpression {
 	updateLocation(fset, (*Node)(c), astNode)
 
 	return (*CallExpression)(c)
+}
+
+func NewCastExpression(fset *token.FileSet, astNode ast.Node) *CastExpression {
+	c, err := env.NewObject("de/fraunhofer/aisec/cpg/graph/statements/expressions/CastExpression")
+	if err != nil {
+		log.Fatal(err)
+
+	}
+
+	updateCode(fset, (*Node)(c), astNode)
+	updateLocation(fset, (*Node)(c), astNode)
+
+	return (*CastExpression)(c)
 }
 
 func NewMemberExpression(fset *token.FileSet, astNode ast.Node) *MemberExpression {
@@ -241,6 +255,14 @@ func (c *CallExpression) SetName(s string) {
 
 func (c *CallExpression) SetFqn(s string) {
 	(*jnigi.ObjectRef)(c).SetField(env, "fqn", NewString(s))
+}
+
+func (c *CastExpression) SetExpression(e *Expression) {
+	(*jnigi.ObjectRef)(c).CallMethod(env, "setExpression", nil, (*jnigi.ObjectRef)(e).Cast("de/fraunhofer/aisec/cpg/graph/statements/expressions/Expression"))
+}
+
+func (c *CastExpression) SetCastType(t *Type) {
+	(*jnigi.ObjectRef)(c).CallMethod(env, "setCastType", nil, t)
 }
 
 func (c *MemberCallExpression) SetName(s string) {

--- a/cpg-language-go/src/main/golang/frontend/handler.go
+++ b/cpg-language-go/src/main/golang/frontend/handler.go
@@ -193,19 +193,25 @@ func (this *GoLanguageFrontend) handleFuncDecl(fset *token.FileSet, funcDecl *as
 
 		var recordType = this.handleType(recv.Type)
 
-		receiver = cpg.NewVariableDeclaration(fset, nil)
+		// The name of the Go receiver is optional. In fact, if the name is not
+		// specified we probably do not need any receiver variable at all,
+		// because the syntax is only there to ensure that this method is part
+		// of the struct, but it is not modifying the receiver.
+		if len(recv.Names) > 0 {
+			receiver = cpg.NewVariableDeclaration(fset, nil)
 
-		// TODO: should we use the FQN here? FQNs are a mess in the CPG...
-		receiver.SetName(recv.Names[0].Name)
-		receiver.SetType(recordType)
+			// TODO: should we use the FQN here? FQNs are a mess in the CPG...
+			receiver.SetName(recv.Names[0].Name)
+			receiver.SetType(recordType)
 
-		err := m.SetReceiver(receiver)
-		if err != nil {
-			log.Fatal(err)
+			err := m.SetReceiver(receiver)
+			if err != nil {
+				log.Fatal(err)
+			}
 		}
 
 		if recordType != nil {
-			var recordName = (*cpg.Node)(recordType).GetName()
+			var recordName = recordType.GetName()
 
 			// TODO: this will only find methods within the current translation unit
 			// this is a limitation that we have for C++ as well
@@ -275,7 +281,7 @@ func (this *GoLanguageFrontend) handleFuncDecl(fset *token.FileSet, funcDecl *as
 		}
 	}
 
-	this.LogDebug("Function has return type %s", (*cpg.Node)(t).GetName())
+	this.LogDebug("Function has return type %s", t.GetName())
 
 	f.SetType(t)
 
@@ -283,13 +289,29 @@ func (this *GoLanguageFrontend) handleFuncDecl(fset *token.FileSet, funcDecl *as
 	// this is a method; however I am not quite sure if this makes sense for
 	// go, since we do not have a 'this', but rather a named receiver
 
-	this.LogDebug("Parsing function body")
-
 	for _, param := range funcDecl.Type.Params.List {
 		p := cpg.NewParamVariableDeclaration(fset, param)
 
-		// TODO: more than one name?
-		p.SetName(param.Names[0].Name)
+		this.LogDebug("Parsing param: %+v", param)
+
+		// Somehow parameters end up having no name sometimes, have not fully understood why.
+		if len(param.Names) > 0 {
+			// TODO: more than one name?
+			var name = param.Names[0].Name
+
+			// If the name is an underscore, it means that the parameter is
+			// unnamed. In order to avoid confusing and some compatibility with
+			// other languages, we are just setting the name to an empty string
+			// in this case.
+			if name == "_" {
+				name = ""
+			}
+
+			p.SetName(name)
+		} else {
+			this.LogError("Some param has no name, which is a bit weird: %+v", param)
+		}
+
 		p.SetType(this.handleType(param.Type))
 
 		// add parameter to scope
@@ -297,6 +319,8 @@ func (this *GoLanguageFrontend) handleFuncDecl(fset *token.FileSet, funcDecl *as
 
 		this.handleComments((*cpg.Node)(p), param)
 	}
+
+	this.LogDebug("Parsing function body of %s", (*cpg.Node)(f).GetName())
 
 	// parse body
 	s := this.handleBlockStmt(fset, funcDecl.Body)
@@ -430,7 +454,7 @@ func (this *GoLanguageFrontend) handleStructTypeSpec(fset *token.FileSet, typeDe
 
 			if field.Names == nil {
 				// retrieve the root type name
-				var typeName = (*cpg.Node)(t.GetRoot()).GetName()
+				var typeName = t.GetRoot().GetName()
 
 				this.LogDebug("Handling embedded field of type %s", typeName)
 
@@ -465,14 +489,25 @@ func (this *GoLanguageFrontend) handleInterfaceTypeSpec(fset *token.FileSet, typ
 
 	if !interfaceType.Incomplete {
 		for _, method := range interfaceType.Methods.List {
-			m := cpg.NewMethodDeclaration(fset, method)
-
 			t := this.handleType(method.Type)
 
-			m.SetType(t)
-			m.SetName(method.Names[0].Name)
+			// Even though this list is called "Methods", it contains all kinds
+			// of things, so we need to proceed with caution. Only if the
+			// "method" actually has a name, we declare a new method
+			// declaration.
+			if len(method.Names) > 0 {
+				m := cpg.NewMethodDeclaration(fset, method)
+				m.SetType(t)
+				m.SetName(method.Names[0].Name)
 
-			scope.AddDeclaration((*cpg.Declaration)(m))
+				scope.AddDeclaration((*cpg.Declaration)(m))
+			} else {
+				this.LogDebug("Adding %s as super class of interface %s", t.GetName(), (*cpg.Node)(r).GetName())
+				// Otherwise, it contains either types or interfaces. For now we
+				// hope that it only has interfaces. We consider embedded
+				// interfaces as sort of super types for this interface.
+				r.AddSuperClass(t)
+			}
 		}
 	}
 
@@ -1102,7 +1137,7 @@ func (this *GoLanguageFrontend) handleCompositeLit(fset *token.FileSet, lit *ast
 	var typ = this.handleType(lit.Type)
 
 	if typ != nil {
-		(*cpg.Node)(c).SetName((*cpg.Node)(typ).GetName())
+		(*cpg.Node)(c).SetName(typ.GetName())
 		(*cpg.Expression)(c).SetType(typ)
 	}
 
@@ -1168,7 +1203,7 @@ func (this *GoLanguageFrontend) handleType(typeExpr ast.Expr) *cpg.Type {
 			log.Fatal(err)
 		}
 
-		this.LogDebug("Pointer to %s", (*cpg.Node)(t).GetName())
+		this.LogDebug("Pointer to %s", t.GetName())
 
 		return t.Reference(i)
 	case *ast.ArrayType:
@@ -1180,7 +1215,7 @@ func (this *GoLanguageFrontend) handleType(typeExpr ast.Expr) *cpg.Type {
 			log.Fatal(err)
 		}
 
-		this.LogDebug("Array of %s", (*cpg.Node)(t).GetName())
+		this.LogDebug("Array of %s", t.GetName())
 
 		return t.Reference(i)
 	case *ast.MapType:
@@ -1190,8 +1225,9 @@ func (this *GoLanguageFrontend) handleType(typeExpr ast.Expr) *cpg.Type {
 		keyType := this.handleType(v.Key)
 		valueType := this.handleType(v.Value)
 
-		(*cpg.ObjectType)(t).AddGeneric(keyType)
-		(*cpg.ObjectType)(t).AddGeneric(valueType)
+		// TODO(oxisto): Find a better way to represent casts
+		(&(cpg.ObjectType{Type: *t})).AddGeneric(keyType)
+		(&(cpg.ObjectType{Type: *t})).AddGeneric(valueType)
 
 		return t
 	case *ast.ChanType:
@@ -1199,7 +1235,7 @@ func (this *GoLanguageFrontend) handleType(typeExpr ast.Expr) *cpg.Type {
 		t := cpg.TypeParser_createFrom("chan", false)
 		chanType := this.handleType(v.Value)
 
-		(*cpg.ObjectType)(t).AddGeneric(chanType)
+		(&(cpg.ObjectType{Type: *t})).AddGeneric(chanType)
 
 		return t
 	case *ast.FuncType:
@@ -1211,7 +1247,7 @@ func (this *GoLanguageFrontend) handleType(typeExpr ast.Expr) *cpg.Type {
 		}
 	}
 
-	return cpg.UnknownType_getUnknown()
+	return &cpg.UnknownType_getUnknown().Type
 }
 
 func (this *GoLanguageFrontend) isBuiltinType(s string) bool {

--- a/cpg-language-go/src/main/golang/frontend/handler.go
+++ b/cpg-language-go/src/main/golang/frontend/handler.go
@@ -661,6 +661,8 @@ func (this *GoLanguageFrontend) handleExpr(fset *token.FileSet, expr ast.Expr) (
 		e = (*cpg.Expression)(this.handleBinaryExpr(fset, v))
 	case *ast.UnaryExpr:
 		e = (*cpg.Expression)(this.handleUnaryExpr(fset, v))
+	case *ast.StarExpr:
+		e = (*cpg.Expression)(this.handleStarExpr(fset, v))
 	case *ast.SelectorExpr:
 		e = (*cpg.Expression)(this.handleSelectorExpr(fset, v))
 	case *ast.KeyValueExpr:
@@ -1019,6 +1021,23 @@ func (this *GoLanguageFrontend) handleUnaryExpr(fset *token.FileSet, unaryExpr *
 	input := this.handleExpr(fset, unaryExpr.X)
 
 	err := u.SetOperatorCode(unaryExpr.Op.String())
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if input != nil {
+		u.SetInput(input)
+	}
+
+	return u
+}
+
+func (this *GoLanguageFrontend) handleStarExpr(fset *token.FileSet, unaryExpr *ast.StarExpr) *cpg.UnaryOperator {
+	u := cpg.NewUnaryOperator(fset, unaryExpr)
+
+	input := this.handleExpr(fset, unaryExpr.X)
+
+	err := u.SetOperatorCode("*")
 	if err != nil {
 		log.Fatal(err)
 	}

--- a/cpg-language-go/src/main/golang/frontend/handler.go
+++ b/cpg-language-go/src/main/golang/frontend/handler.go
@@ -671,6 +671,10 @@ func (this *GoLanguageFrontend) handleExpr(fset *token.FileSet, expr ast.Expr) (
 		e = (*cpg.Expression)(this.handleCompositeLit(fset, v))
 	case *ast.Ident:
 		e = (*cpg.Expression)(this.handleIdent(fset, v))
+	case *ast.TypeAssertExpr:
+		e = (*cpg.Expression)(this.handleTypeAssertExpr(fset, v))
+	case *ast.ParenExpr:
+		e = this.handleExpr(fset, v.X)
 	default:
 		this.LogError("Could not parse expression of type %T: %+v", v, v)
 		// TODO: return an error instead?
@@ -1177,6 +1181,21 @@ func (this *GoLanguageFrontend) handleIdent(fset *token.FileSet, ident *ast.Iden
 	ref.SetName(ident.Name)
 
 	return ref
+}
+
+func (this *GoLanguageFrontend) handleTypeAssertExpr(fset *token.FileSet, assert *ast.TypeAssertExpr) *cpg.CastExpression {
+	cast := cpg.NewCastExpression(fset, assert)
+
+	// Parse the inner expression
+	expr := this.handleExpr(fset, assert.X)
+
+	// Parse the type
+	typ := this.handleType(assert.Type)
+
+	cast.SetExpression(expr)
+	cast.SetCastType(typ)
+
+	return cast
 }
 
 func (this *GoLanguageFrontend) handleType(typeExpr ast.Expr) *cpg.Type {

--- a/cpg-language-go/src/main/golang/types.go
+++ b/cpg-language-go/src/main/golang/types.go
@@ -36,11 +36,15 @@ import (
 
 var env *jnigi.Env
 
-type Type jnigi.ObjectRef
+type Type struct{ *jnigi.ObjectRef }
 
 func (t *Type) ConvertToGo(o *jnigi.ObjectRef) error {
-	*t = (Type)(*o)
+	t.ObjectRef = o
 	return nil
+}
+
+func (t *Type) ConvertToJava() (obj *jnigi.ObjectRef, err error) {
+	return t.ObjectRef, nil
 }
 
 func (*Type) GetClassName() string {
@@ -51,19 +55,25 @@ func (*Type) IsArray() bool {
 	return false
 }
 
-type ObjectType jnigi.ObjectRef
+func (t *Type) GetName() string {
+	// A little bit hacky until we also convert node to a struct
+	return (*Node)(t.ObjectRef).GetName()
+}
 
-func (t *ObjectType) ConvertToGo(o *jnigi.ObjectRef) error {
-	*t = (ObjectType)(*o)
-	return nil
+type ObjectType struct {
+	Type
 }
 
 func (*ObjectType) GetClassName() string {
 	return "de/fraunhofer/aisec/cpg/graph/types/ObjectType"
 }
 
-func (*ObjectType) IsArray() bool {
-	return false
+type UnknownType struct {
+	Type
+}
+
+func (*UnknownType) GetClassName() string {
+	return "de/fraunhofer/aisec/cpg/graph/types/UnknownType"
 }
 
 type HasType jnigi.ObjectRef
@@ -83,8 +93,8 @@ func TypeParser_createFrom(s string, resolveAlias bool) *Type {
 	return &t
 }
 
-func UnknownType_getUnknown() *Type {
-	var t Type
+func UnknownType_getUnknown() *UnknownType {
+	var t UnknownType
 	err := env.CallStaticMethod("de/fraunhofer/aisec/cpg/graph/types/UnknownType", "getUnknownType", &t)
 	if err != nil {
 		log.Fatal(err)
@@ -94,19 +104,19 @@ func UnknownType_getUnknown() *Type {
 	return &t
 }
 
-func (h *Type) GetRoot() *Type {
-	var t Type
-	err := (*jnigi.ObjectRef)(h).CallMethod(env, "getRoot", &t)
+func (t *Type) GetRoot() *Type {
+	var root Type
+	err := t.CallMethod(env, "getRoot", &root)
 	if err != nil {
 		log.Fatal(err)
 	}
 
-	return &t
+	return &root
 }
 
 func (t *Type) Reference(o *jnigi.ObjectRef) *Type {
 	var refType Type
-	err := (*jnigi.ObjectRef)(t).CallMethod(env, "reference", &refType, (*jnigi.ObjectRef)(o).Cast("de/fraunhofer/aisec/cpg/graph/types/PointerType$PointerOrigin"))
+	err := t.CallMethod(env, "reference", &refType, (*jnigi.ObjectRef)(o).Cast("de/fraunhofer/aisec/cpg/graph/types/PointerType$PointerOrigin"))
 
 	if err != nil {
 		log.Fatal(err)
@@ -117,7 +127,7 @@ func (t *Type) Reference(o *jnigi.ObjectRef) *Type {
 
 func (h *HasType) SetType(t *Type) {
 	if t != nil {
-		(*jnigi.ObjectRef)(h).CallMethod(env, "setType", nil, (*jnigi.ObjectRef)(t).Cast("de/fraunhofer/aisec/cpg/graph/types/Type"))
+		(*jnigi.ObjectRef)(h).CallMethod(env, "setType", nil, t.Cast("de/fraunhofer/aisec/cpg/graph/types/Type"))
 	}
 }
 
@@ -134,8 +144,8 @@ func (h *HasType) GetType() *Type {
 func (t *ObjectType) AddGeneric(g *Type) {
 	// Stupid workaround, since casting does not work. See
 	// https://github.com/timob/jnigi/issues/60
-	var objType = jnigi.WrapJObject(uintptr((*jnigi.ObjectRef)(t).JObject()), "de/fraunhofer/aisec/cpg/graph/types/ObjectType", false)
-	err := objType.CallMethod(env, "addGeneric", nil, (*jnigi.ObjectRef)(g).Cast("de/fraunhofer/aisec/cpg/graph/types/Type"))
+	var objType = jnigi.WrapJObject(uintptr(t.JObject()), "de/fraunhofer/aisec/cpg/graph/types/ObjectType", false)
+	err := objType.CallMethod(env, "addGeneric", nil, g.Cast("de/fraunhofer/aisec/cpg/graph/types/Type"))
 	if err != nil {
 		log.Fatal(err)
 	}

--- a/cpg-language-go/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/golang/DeclarationTest.kt
+++ b/cpg-language-go/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/golang/DeclarationTest.kt
@@ -1,0 +1,125 @@
+/*
+ * Copyright (c) 2022, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.frontends.golang
+
+import de.fraunhofer.aisec.cpg.ExperimentalGolang
+import de.fraunhofer.aisec.cpg.TestUtils
+import de.fraunhofer.aisec.cpg.graph.byNameOrNull
+import de.fraunhofer.aisec.cpg.graph.declarations.FunctionDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.MethodDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.NamespaceDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.RecordDeclaration
+import java.nio.file.Path
+import kotlin.test.*
+
+@OptIn(ExperimentalGolang::class)
+class DeclarationTest {
+    @Test
+    fun testUnnamedReceiver() {
+        val topLevel = Path.of("src", "test", "resources", "golang")
+        val tu =
+            TestUtils.analyzeAndGetFirstTU(
+                listOf(topLevel.resolve("unnamed.go").toFile()),
+                topLevel,
+                true
+            ) {
+                it.registerLanguage(
+                    GoLanguageFrontend::class.java,
+                    GoLanguageFrontend.GOLANG_EXTENSIONS
+                )
+            }
+        assertNotNull(tu)
+
+        val main = tu.byNameOrNull<NamespaceDeclaration>("main")
+        assertNotNull(main)
+
+        val myStruct = main.byNameOrNull<RecordDeclaration>("main.MyStruct")
+        assertNotNull(myStruct)
+
+        // Receiver should be null since its unnamed
+        val myFunc = myStruct.byNameOrNull<MethodDeclaration>("MyFunc")
+        assertNotNull(myFunc)
+        assertNull(myFunc.receiver)
+    }
+
+    @Test
+    fun testUnnamedParameter() {
+        val topLevel = Path.of("src", "test", "resources", "golang")
+        val tu =
+            TestUtils.analyzeAndGetFirstTU(
+                listOf(topLevel.resolve("unnamed.go").toFile()),
+                topLevel,
+                true
+            ) {
+                it.registerLanguage(
+                    GoLanguageFrontend::class.java,
+                    GoLanguageFrontend.GOLANG_EXTENSIONS
+                )
+            }
+        assertNotNull(tu)
+
+        val main = tu.byNameOrNull<NamespaceDeclaration>("main")
+        assertNotNull(main)
+
+        // Parameter should be there but not have a name
+        val myGlobalFunc = main.byNameOrNull<FunctionDeclaration>("MyGlobalFunc")
+        assertNotNull(myGlobalFunc)
+
+        val param = myGlobalFunc.parameters.firstOrNull()
+        assertNotNull(param)
+        assertEquals("", param.name)
+    }
+
+    @Test
+    fun testEmbeddedInterface() {
+        val topLevel = Path.of("src", "test", "resources", "golang")
+        val tu =
+            TestUtils.analyzeAndGetFirstTU(
+                listOf(topLevel.resolve("embed.go").toFile()),
+                topLevel,
+                true
+            ) {
+                it.registerLanguage(
+                    GoLanguageFrontend::class.java,
+                    GoLanguageFrontend.GOLANG_EXTENSIONS
+                )
+            }
+        assertNotNull(tu)
+
+        val main = tu.byNameOrNull<NamespaceDeclaration>("main")
+        assertNotNull(main)
+
+        val myInterface = main.byNameOrNull<RecordDeclaration>("main.MyInterface")
+        assertNotNull(myInterface)
+
+        val myOtherInterface = main.byNameOrNull<RecordDeclaration>("main.MyOtherInterface")
+        assertNotNull(myOtherInterface)
+
+        // MyOtherInterface should be in the superClasses and superTypeDeclarations of MyInterface, since it is embedded and thus MyInterface "extends" it
+        assertContains(myInterface.superTypeDeclarations, myOtherInterface)
+        assertTrue(myInterface.superClasses.any { it.name == myOtherInterface.name })
+    }
+}

--- a/cpg-language-go/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/golang/DeclarationTest.kt
+++ b/cpg-language-go/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/golang/DeclarationTest.kt
@@ -118,7 +118,8 @@ class DeclarationTest {
         val myOtherInterface = main.byNameOrNull<RecordDeclaration>("main.MyOtherInterface")
         assertNotNull(myOtherInterface)
 
-        // MyOtherInterface should be in the superClasses and superTypeDeclarations of MyInterface, since it is embedded and thus MyInterface "extends" it
+        // MyOtherInterface should be in the superClasses and superTypeDeclarations of MyInterface,
+        // since it is embedded and thus MyInterface "extends" it
         assertContains(myInterface.superTypeDeclarations, myOtherInterface)
         assertTrue(myInterface.superClasses.any { it.name == myOtherInterface.name })
     }

--- a/cpg-language-go/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/golang/ExpressionTest.kt
+++ b/cpg-language-go/src/test/kotlin/de/fraunhofer/aisec/cpg/frontends/golang/ExpressionTest.kt
@@ -1,0 +1,83 @@
+/*
+ * Copyright (c) 2022, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.frontends.golang
+
+import de.fraunhofer.aisec.cpg.ExperimentalGolang
+import de.fraunhofer.aisec.cpg.TestUtils
+import de.fraunhofer.aisec.cpg.graph.bodyOrNull
+import de.fraunhofer.aisec.cpg.graph.byNameOrNull
+import de.fraunhofer.aisec.cpg.graph.declarations.FunctionDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.NamespaceDeclaration
+import de.fraunhofer.aisec.cpg.graph.declarations.VariableDeclaration
+import de.fraunhofer.aisec.cpg.graph.statements.DeclarationStatement
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.CastExpression
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.DeclaredReferenceExpression
+import java.nio.file.Path
+import kotlin.test.assertEquals
+import kotlin.test.assertNotNull
+import kotlin.test.assertSame
+import org.junit.jupiter.api.Test
+
+class ExpressionTest {
+    @OptIn(ExperimentalGolang::class)
+    @Test
+    fun testTypeAssert() {
+        val topLevel = Path.of("src", "test", "resources", "golang")
+        val tu =
+            TestUtils.analyzeAndGetFirstTU(
+                listOf(topLevel.resolve("type_assert.go").toFile()),
+                topLevel,
+                true
+            ) {
+                it.registerLanguage(
+                    GoLanguageFrontend::class.java,
+                    GoLanguageFrontend.GOLANG_EXTENSIONS
+                )
+            }
+        assertNotNull(tu)
+
+        val main = tu.byNameOrNull<NamespaceDeclaration>("main")
+        assertNotNull(main)
+
+        val mainFunc = main.byNameOrNull<FunctionDeclaration>("main")
+        assertNotNull(mainFunc)
+
+        val f =
+            (mainFunc.bodyOrNull<DeclarationStatement>(0))?.singleDeclaration
+                as? VariableDeclaration
+        assertNotNull(f)
+
+        val s =
+            (mainFunc.bodyOrNull<DeclarationStatement>(1))?.singleDeclaration
+                as? VariableDeclaration
+        assertNotNull(s)
+
+        val cast = s.initializer as? CastExpression
+        assertNotNull(cast)
+        assertEquals("main.MyStruct", cast.castType.name)
+        assertSame(f, (cast.expression as? DeclaredReferenceExpression)?.refersTo)
+    }
+}

--- a/cpg-language-go/src/test/resources/golang/embed.go
+++ b/cpg-language-go/src/test/resources/golang/embed.go
@@ -1,0 +1,7 @@
+package main
+
+type MyInterface interface {
+    MyOtherInterface
+}
+
+type MyOtherInterface interface {}

--- a/cpg-language-go/src/test/resources/golang/type_assert.go
+++ b/cpg-language-go/src/test/resources/golang/type_assert.go
@@ -1,0 +1,16 @@
+package main
+
+import "fmt"
+
+type MyStruct struct {}
+type MyInterface interface {
+    MyFunc()
+}
+func (MyStruct) MyFunc() {}
+
+func main () {
+    var f MyInterface = MyStruct{}
+    var s = f.(MyStruct)
+
+    fmt.Printf("%+v", s)
+}

--- a/cpg-language-go/src/test/resources/golang/unnamed.go
+++ b/cpg-language-go/src/test/resources/golang/unnamed.go
@@ -1,0 +1,13 @@
+package main
+
+import "context"
+
+type MyStruct struct{}
+
+func (MyStruct) MyFunc() int {
+	return 1
+}
+
+func MyGlobalFunc(_ context.Context) int {
+	return 2
+}

